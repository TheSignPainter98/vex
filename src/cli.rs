use std::fmt::Display;

use camino::Utf8PathBuf;
use clap::{
    builder::{StringValueParser, TypedValueParser},
    ArgAction, Parser, Subcommand,
};

#[derive(Debug, Parser)]
#[command(
    author,
    version,
    about,
    disable_help_flag = true,
    disable_version_flag = true
)]
// #[warn(missing_docs)]
pub struct Args {
    #[command(subcommand)]
    pub command: Option<Command>,

    #[arg(short, value_name="level", action=ArgAction::Count, value_name="level", global=true)]
    pub verbosity_level: u8,

    /// Print help information, use `--help` for more detail
    #[arg(short, long, action=ArgAction::Help, global=true)]
    pub help: Option<bool>,
}

#[derive(Debug, Subcommand)]
pub enum Command {
    #[command(name = "languages")]
    ListLanguages,

    #[command(name = "lints")]
    ListLints,

    Check(CheckCmd),

    Dump(DumpCmd),

    Init,
}

impl Default for Command {
    fn default() -> Self {
        Self::Check(CheckCmd::default())
    }
}

#[derive(Debug, Default, Parser)]
pub struct CheckCmd {
    // #[arg(long, default_value_t = MaxConcurrentFileLimit::default(), value_parser = MaxConcurrentFileLimit::parser())]
    // pub max_concurrent_files: MaxConcurrentFileLimit,
    //
    // TODO(kcza): use me!
    #[arg(long, default_value_t = MaxProblems::default(), value_parser = MaxProblems::parser())]
    pub max_problems: MaxProblems,
}

// #[derive(Clone, Debug)]
// pub struct MaxConcurrentFileLimit(pub u32);
//
// impl MaxConcurrentFileLimit {
//     fn parser() -> impl TypedValueParser {
//         StringValueParser::new().try_map(|s| Ok::<_, anyhow::Error>(Self(s.parse()?)))
//     }
// }
//
// impl Default for MaxConcurrentFileLimit {
//     fn default() -> Self {
//         Self(16)
//     }
// }
//
// impl Display for MaxConcurrentFileLimit {
//     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//         self.0.fmt(f)
//     }
// }

#[derive(Clone, Debug)]
pub enum MaxProblems {
    Unlimited,
    Limited(u32),
}

impl MaxProblems {
    fn parser() -> impl TypedValueParser {
        StringValueParser::new().try_map(|s| {
            if s.to_lowercase() == "unlimited" {
                return Ok(Self::Unlimited);
            }

            let max: u32 = s.parse()?;
            Ok::<_, anyhow::Error>(Self::Limited(max))
        })
    }

    #[allow(unused)]
    pub fn is_exceeded_by(&self, to_check: usize) -> bool {
        match self {
            Self::Unlimited => false,
            Self::Limited(lim) => to_check >= *lim as usize,
        }
    }
}

impl Default for MaxProblems {
    fn default() -> Self {
        Self::Limited(100)
    }
}

impl Display for MaxProblems {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Unlimited => write!(f, "unlimited"),
            Self::Limited(l) => l.fmt(f),
        }
    }
}

#[derive(Debug, Default, Parser)]
pub struct DumpCmd {
    pub path: Utf8PathBuf,
}
