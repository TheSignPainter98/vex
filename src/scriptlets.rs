pub mod action;
pub mod app_object;
pub mod event;
mod extra_data;
mod observers;
mod print_handler;
mod scriptlet;
mod store;

pub use observers::{Observer, ScriptletObserverData};
pub use store::{PreinitingStore, VexingStore};
