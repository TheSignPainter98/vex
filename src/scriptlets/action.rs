use allocative::Allocative;
use dupe::Dupe;

use crate::scriptlets::event::EventKind;

#[derive(Copy, Clone, Debug, PartialEq, Eq, Allocative, Dupe)]
pub enum Action {
    Preiniting,
    Initing,
    Vexing(EventKind),
}

impl Action {
    pub fn pretty_name(&self) -> &'static str {
        match self {
            Self::Preiniting => "preiniting",
            Self::Initing => "initing",
            Self::Vexing(e) => e.pretty_name(),
        }
    }
}
