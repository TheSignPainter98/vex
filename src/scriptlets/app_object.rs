use std::fmt::Display;

use allocative::Allocative;
use starlark::{
    environment::{Methods, MethodsBuilder, MethodsStatic},
    eval::Evaluator,
    starlark_module,
    values::Value,
    values::{none::NoneType, NoSerialize, ProvidesStaticType, StarlarkValue},
};
use starlark_derive::starlark_value;

use crate::{
    error::Error,
    scriptlets::{
        action::Action,
        event::EventType,
        extra_data::{HandlerDataBuilder, InvocationData},
    },
};

#[derive(Debug, PartialEq, Eq, ProvidesStaticType, NoSerialize, Allocative)]
pub struct AppObject;

impl AppObject {
    pub const NAME: &'static str = "vex";

    #[starlark_module]
    fn methods(builder: &mut MethodsBuilder) {
        fn language<'v>(
            #[starlark(this)] _this: Value<'v>,
            lang: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.language", &[Action::Initing])?;

            HandlerDataBuilder::get_from(eval.module()).set_language(lang.into());

            Ok(NoneType)
        }

        fn query<'v>(
            #[starlark(this)] _this: Value<'v>,
            query: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.query", &[Action::Initing])?;

            // TODO(kcza): attach the id in errors somewhere?
            HandlerDataBuilder::get_from(eval.module()).set_query(query.into());

            Ok(NoneType)
        }

        fn observe<'v>(
            #[starlark(this)] _this: Value<'v>,
            event: &str,
            handler: Value<'v>,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.observe", &[Action::Initing])?;

            let event = event.parse()?;
            HandlerDataBuilder::get_from(eval.module()).add_observer(event, handler);

            Ok(NoneType)
        }

        fn warn<'v>(
            #[starlark(this)] _this: Value<'v>,
            _msg: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(
                eval,
                "vex.warn",
                &[
                    Action::Vexing(EventType::Start),
                    Action::Vexing(EventType::Match),
                    Action::Vexing(EventType::EoF),
                    Action::Vexing(EventType::End),
                ],
            )?;

            todo!();
            // Ok(NoneType)
        }
    }

    fn check_attr_available(
        eval: &Evaluator<'_, '_>,
        attr_path: &'static str,
        available_actions: &[Action],
    ) -> anyhow::Result<()> {
        let curr_action = InvocationData::get_from(eval).action();
        if !available_actions.contains(&curr_action) {
            return Err(Error::Unavailable {
                what: attr_path,
                action: curr_action,
            }
            .into());
        }
        Ok(())
    }
}

starlark::starlark_simple_value!(AppObject);
#[starlark_value(type = "vex")]
impl<'v> StarlarkValue<'v> for AppObject {
    fn get_methods() -> Option<&'static Methods> {
        static RES: MethodsStatic = MethodsStatic::new();
        RES.methods(AppObject::methods)
    }
}

impl Display for AppObject {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        Self::NAME.fmt(f)
    }
}
