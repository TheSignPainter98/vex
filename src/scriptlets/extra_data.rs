use std::{cell::RefCell, fmt::Display};

use allocative::Allocative;
use camino::Utf8Path;
use dupe::Dupe;
use starlark::{
    environment::{FrozenModule, Module},
    eval::Evaluator,
    values::{
        AllocFrozenValue, AllocValue, Demand, Freeze, Freezer, FrozenHeap, FrozenValue, Heap,
        ProvidesStaticType, StarlarkValue, Trace, Tracer, Value, ValueLike,
    },
};
use starlark_derive::{starlark_value, NoSerialize};
use tree_sitter::Query;

use crate::{
    error::Error,
    scriptlets::{action::Action, event::EventType},
    supported_language::SupportedLanguage,
};

#[derive(Debug, ProvidesStaticType, NoSerialize, Allocative)]
pub struct EvaluatorData {
    action: Action,
}

impl EvaluatorData {
    pub fn new(action: Action) -> Self {
        Self { action }
    }

    pub fn insert_into<'e>(&'e self, eval: &mut Evaluator<'_, 'e>) {
        eval.extra = Some(self);
    }

    pub fn get_from<'e>(eval: &Evaluator<'_, 'e>) -> &'e Self {
        eval.extra
            .as_ref()
            .expect("Evaluator extra not set")
            .downcast_ref()
            .expect("Evaluator extra has wrong type")
    }

    pub fn action(&self) -> Action {
        self.action
    }
}

starlark::starlark_simple_value!(EvaluatorData);
#[starlark_value(type = "EvaluatorData")]
impl<'v> StarlarkValue<'v> for EvaluatorData {}

impl Display for EvaluatorData {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", EvaluatorData::TYPE)
    }
}

#[derive(Debug, Trace, Default, ProvidesStaticType, NoSerialize, Allocative)]
pub struct HandlerDataBuilder<'v> {
    #[allocative(skip)]
    pub lang: RefCell<Option<RawSupportedLanguage<'v>>>,
    #[allocative(skip)]
    pub query: RefCell<Option<RawQuery<'v>>>,
    pub on_start: RefCell<Vec<Value<'v>>>,
    pub on_match: RefCell<Vec<Value<'v>>>,
    pub on_eof: RefCell<Vec<Value<'v>>>,
    pub on_end: RefCell<Vec<Value<'v>>>,
}

impl<'v> HandlerDataBuilder<'v> {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert_into(self, module: &'v Module) {
        module.set_extra_value(module.heap().alloc(self))
    }

    pub fn get_from(module: &'v Module) -> &'v Self {
        module
            .extra_value()
            .as_ref()
            .expect("Module extra not set")
            .request_value()
            .expect("Module extra has wrong type")
    }

    pub fn set_language(&self, language: RawSupportedLanguage<'v>) {
        *self.lang.borrow_mut() = Some(language);
    }

    pub fn set_query(&self, query: RawQuery<'v>) {
        *self.query.borrow_mut() = Some(query);
    }

    pub fn add_observer(&self, event: EventType, handler: Value<'v>) {
        match event {
            EventType::Start => self.on_start.borrow_mut().push(handler),
            EventType::Match => self.on_match.borrow_mut().push(handler),
            EventType::EoF => self.on_eof.borrow_mut().push(handler),
            EventType::End => self.on_end.borrow_mut().push(handler),
        }
    }
}

#[starlark_value(type = "HandlerDataBuilder")]
impl<'v> StarlarkValue<'v> for HandlerDataBuilder<'v> {
    fn provide(&'v self, demand: &mut Demand<'_, 'v>) {
        demand.provide_value(self)
    }
}

impl<'v> Freeze for HandlerDataBuilder<'v> {
    type Frozen = FrozenHandlerDataBuilder;

    fn freeze(self, freezer: &Freezer) -> anyhow::Result<Self::Frozen> {
        let HandlerDataBuilder {
            lang,
            query,
            on_start,
            on_match,
            on_eof,
            on_end,
        } = self;
        let lang = lang.into_inner().map(|e| e.to_string());
        let query = query.into_inner().map(|e| e.to_string());
        let on_start = on_start
            .into_inner()
            .into_iter()
            .map(|v| v.freeze(freezer))
            .collect::<anyhow::Result<_>>()?;
        let on_match = on_match
            .into_inner()
            .into_iter()
            .map(|v| v.freeze(freezer))
            .collect::<anyhow::Result<_>>()?;
        let on_eof = on_eof
            .into_inner()
            .into_iter()
            .map(|v| v.freeze(freezer))
            .collect::<anyhow::Result<_>>()?;
        let on_end = on_end
            .into_inner()
            .into_iter()
            .map(|v| v.freeze(freezer))
            .collect::<anyhow::Result<_>>()?;
        Ok(FrozenHandlerDataBuilder {
            lang,
            query,
            on_start,
            on_match,
            on_eof,
            on_end,
        })
    }
}

impl<'v> AllocValue<'v> for HandlerDataBuilder<'v> {
    #[inline]
    fn alloc_value(self, heap: &'v Heap) -> Value<'v> {
        heap.alloc_complex(self)
    }
}

impl<'v> Display for HandlerDataBuilder<'v> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Self::TYPE.fmt(f)
    }
}

pub type RawSupportedLanguage<'v> = RawStr<'v>;
pub type RawQuery<'v> = RawStr<'v>;

/// Wrapper type to allow implementation of certain traits on &str.
#[derive(Debug)]
pub struct RawStr<'v>(&'v str);

unsafe impl<'v> Trace<'v> for RawStr<'v> {
    fn trace(&mut self, _tracer: &Tracer<'v>) {}
}

impl<'v> From<&'v str> for RawStr<'v> {
    fn from(value: &'v str) -> Self {
        Self(value)
    }
}

impl<'v> Display for RawStr<'v> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug, ProvidesStaticType, NoSerialize, Allocative)]
pub struct FrozenHandlerDataBuilder {
    pub lang: Option<String>,
    pub query: Option<String>,
    pub on_start: Vec<FrozenValue>,
    pub on_match: Vec<FrozenValue>,
    pub on_eof: Vec<FrozenValue>,
    pub on_end: Vec<FrozenValue>,
}

impl FrozenHandlerDataBuilder {
    pub fn get_from(frozen_module: &FrozenModule) -> &Self {
        frozen_module
            .extra_value()
            .as_ref()
            .expect("FrozenModule extra not set")
            .downcast_ref()
            .expect("FrozenModule extra has wrong type")
    }

    pub fn build(&self, path: &Utf8Path) -> anyhow::Result<HandlerData> {
        let Self {
            lang,
            query,
            on_start,
            on_match,
            on_eof,
            on_end,
        } = self;

        if on_start.len() + on_match.len() + on_end.len() + on_end.len() == 0 {
            return Err(Error::NoCallbacks(path.to_owned()).into());
        }

        let lang = {
            let Some(lang) = lang else {
                return Err(Error::NoLanguage(path.to_owned()).into());
            };
            lang.parse::<SupportedLanguage>()?
        };
        let query = {
            let Some(query) = query else {
                return Err(Error::NoQuery(path.to_owned()).into());
            };
            Query::new(lang.ts_language(), &query)?
        };
        let on_start = on_start.iter().map(Dupe::dupe).collect();
        let on_match = on_match.iter().map(Dupe::dupe).collect();
        let on_eof = on_eof.iter().map(Dupe::dupe).collect();
        let on_end = on_end.iter().map(Dupe::dupe).collect();
        Ok(HandlerData {
            lang,
            query,
            on_start,
            on_match,
            on_eof,
            on_end,
        })
    }
}

#[starlark_value(type = "HandlerData")]
impl<'v> StarlarkValue<'v> for FrozenHandlerDataBuilder {
    type Canonical = HandlerDataBuilder<'v>;

    fn provide(&'v self, demand: &mut Demand<'_, 'v>) {
        demand.provide_value(self)
    }
}

impl AllocFrozenValue for FrozenHandlerDataBuilder {
    #[inline]
    fn alloc_frozen_value(self, heap: &FrozenHeap) -> FrozenValue {
        heap.alloc_simple(self)
    }
}

impl Display for FrozenHandlerDataBuilder {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Self::TYPE.fmt(f)
    }
}

#[derive(Debug)]
pub struct HandlerData {
    pub lang: SupportedLanguage,
    pub query: Query,
    pub on_start: Vec<FrozenValue>,
    pub on_match: Vec<FrozenValue>,
    pub on_eof: Vec<FrozenValue>,
    pub on_end: Vec<FrozenValue>,
}
