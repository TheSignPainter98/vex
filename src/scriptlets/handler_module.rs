use std::ops::Deref;

use starlark::{environment::Module, values::FrozenHeap};

use crate::{
    result::Result,
    scriptlets::{
        extra_data::{RetainedData, UnfrozenRetainedData},
        Intents,
    },
};

pub struct HandlerModule {
    module: Module,
}

impl HandlerModule {
    pub fn new() -> Self {
        let module = Module::new();
        let ret_data = UnfrozenRetainedData::new();
        ret_data.insert_into(&module);

        Self { module }
    }

    pub fn intent_count(&self) -> usize {
        let ret_data = UnfrozenRetainedData::get_from(&self.module);
        ret_data.intent_count()
    }

    pub fn into_intents_on(self, frozen_heap: &FrozenHeap) -> Result<Intents> {
        let Self { module, .. } = self;
        let module = module.freeze()?;
        frozen_heap.add_reference(module.frozen_heap());

        let ret_data = RetainedData::get_from(&module);
        Ok(ret_data.intents().clone())
    }

    pub fn into_module(self) -> Module {
        self.module
    }
}

impl Deref for HandlerModule {
    type Target = Module;

    fn deref(&self) -> &Self::Target {
        &self.module
    }
}
