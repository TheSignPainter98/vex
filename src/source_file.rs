use std::{fs, sync::Arc};

use anyhow::Context;
use camino::Utf8Path;
use log::{log_enabled, trace};
use tree_sitter::{Parser, Tree};

use crate::supported_language::SupportedLanguage;

pub struct SourceFile {
    pub path: Arc<Utf8Path>,
    pub content: String,
    pub tree: Tree,
    pub lang: SupportedLanguage,
}

impl SourceFile {
    pub fn load_if_supported(path: Arc<Utf8Path>) -> Option<anyhow::Result<Self>> {
        let Some(extension) = path.extension() else {
            if log_enabled!(log::Level::Trace) {
                trace!("ignoring {path} (no file extension)");
            }
            return None;
        };
        let Some(lang) = SupportedLanguage::try_from_extension(extension) else {
            if log_enabled!(log::Level::Trace) {
                trace!("ignoring {path} (no known language)");
            }
            return None;
        };
        Some(Self::load(path, lang))
    }

    fn load(path: Arc<Utf8Path>, lang: SupportedLanguage) -> anyhow::Result<Self> {
        let content = fs::read_to_string(path.as_ref())?;
        let tree = {
            let mut parser = Parser::new();
            parser
                .set_language(lang.ts_language())
                .with_context(|| format!("failed to load {} grammar", lang.name()))?;
            parser
                .parse(&content, None)
                .with_context(|| format!("failed to parse {path}"))?
        };
        Ok(Self {
            path,
            content,
            lang,
            tree,
        })
    }
}
