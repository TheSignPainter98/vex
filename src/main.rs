#![deny(missing_debug_implementations)]

#[cfg(test)]
#[macro_use]
extern crate pretty_assertions;

mod associations;
mod cli;
mod context;
mod error;
mod ignore_markers;
mod irritation;
mod logger;
mod parse;
mod plural;
mod query;
mod result;
mod scriptlets;
mod source_file;
mod source_path;
mod suggestion;
mod supported_language;
mod trigger;
mod verbosity;
mod vex;

#[cfg(test)]
mod vextest;

use std::{env, fs, process::ExitCode};

use camino::{Utf8Path, Utf8PathBuf};
use cli::{InitCmd, ListCmd, MaxProblems, ToList};
use dupe::Dupe;
use indoc::printdoc;
use lazy_static::lazy_static;
use log::{info, log_enabled, trace, warn};
use owo_colors::{OwoColorize, Stream, Style};
use scriptlets::{
    action::Action, event::EventKind, handler_module::HandlerModule, Observable, ObserveOptions,
};
use source_file::SourceFile;
use strum::IntoEnumIterator;
use tree_sitter::QueryCursor;

use crate::{
    cli::{Args, CheckCmd, Command},
    context::{Context, EXAMPLE_VEX_FILE},
    error::{Error, IOAction},
    irritation::Irritation,
    plural::Plural,
    result::Result,
    scriptlets::{
        event::{MatchEvent, OpenFileEvent, OpenProjectEvent},
        intents::Intent,
        query_cache::QueryCache,
        query_captures::QueryCaptures,
        PreinitOptions, PreinitingStore, VexingStore,
    },
    source_path::{PrettyPath, SourcePath},
    supported_language::SupportedLanguage,
    trigger::FilePattern,
    verbosity::Verbosity,
};

// TODO(kcza): move the subcommands to separate files
fn main() -> ExitCode {
    match run() {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{e}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<ExitCode> {
    let args = Args::parse();
    logger::init(Verbosity::try_from(args.verbosity_level)?)?;

    match args.command {
        Command::Check(cmd_args) => check(cmd_args),
        Command::List(list_args) => list(list_args),
        Command::Init(init_args) => init(init_args),
        Command::Parse(parse_args) => parse::parse(parse_args),
    }?;

    Ok(logger::exit_code())
}

fn list(list_args: ListCmd) -> Result<()> {
    match list_args.what {
        ToList::Checks => {
            let ctx = Context::acquire()?;
            let store = PreinitingStore::new(&ctx)?.preinit(PreinitOptions::default())?;
            store
                .vexes()
                .map(|vex| &vex.vex_id)
                .for_each(|id| println!("{}", id));
        }
        ToList::Languages => SupportedLanguage::iter().for_each(|lang| println!("{}", lang)),
    }
    Ok(())
}

lazy_static! {
    static ref SUCCESS_STYLE: Style = Style::new().green().bold();
}

fn check(cmd_args: CheckCmd) -> Result<()> {
    let ctx = Context::acquire()?;
    let store = {
        let preinit_opts = PreinitOptions {
            lenient: cmd_args.lenient,
        };
        PreinitingStore::new(&ctx)?.preinit(preinit_opts)?.init()?
    };

    let RunData {
        irritations,
        num_files_scanned,
    } = vex(&ctx, &store, cmd_args.max_problems)?;
    irritations.iter().for_each(|irr| println!("{irr}"));
    if log_enabled!(log::Level::Info) {
        info!(
            "scanned {}",
            Plural::new(num_files_scanned, "file", "files"),
        );
    }
    if !irritations.is_empty() {
        warn!(
            "found {}",
            Plural::new(irritations.len(), "problem", "problems"),
        );
    } else {
        println!(
            "{}: no problems found",
            "success".if_supports_color(Stream::Stdout, |text| text.style(*SUCCESS_STYLE))
        );
    }

    Ok(())
}

#[derive(Debug)]
struct RunData {
    irritations: Vec<Irritation>,
    num_files_scanned: usize,
}

impl RunData {
    #[cfg(test)]
    fn into_irritations(self) -> Vec<Irritation> {
        self.irritations
    }
}

fn vex(ctx: &Context, store: &VexingStore, max_problems: MaxProblems) -> Result<RunData> {
    let files = {
        let mut paths = Vec::new();
        let ignores = ctx
            .ignores
            .clone()
            .into_inner()
            .into_iter()
            .map(|ignore| ignore.compile())
            .collect::<Result<Vec<_>>>()?;
        let allows = ctx
            .allows
            .clone()
            .into_iter()
            .map(|allow| allow.compile())
            .collect::<Result<Vec<_>>>()?;
        walkdir(
            ctx,
            ctx.project_root.as_ref(),
            &ignores,
            &allows,
            &mut paths,
        )?;

        let associations = ctx.associations()?;
        paths
            .into_iter()
            .map(|p| SourcePath::new(&p, &ctx.project_root))
            .map(|p| {
                let language = associations.get_language(&p)?;
                Ok(SourceFile::new(p, language))
            })
            .collect::<Result<Vec<_>>>()?
    };

    let project_queries_hint = store.project_queries_hint();
    let file_queries_hint = store.file_queries_hint();

    let query_cache = QueryCache::with_capacity(project_queries_hint + file_queries_hint);

    let mut irritations = vec![];
    let frozen_heap = store.frozen_heap();
    let project_queries = {
        let mut project_queries = Vec::with_capacity(project_queries_hint);

        let event = OpenProjectEvent::new(ctx.project_root.dupe());
        let handler_module = HandlerModule::new();
        let observe_opts = ObserveOptions {
            action: Action::Vexing(event.kind()),
            query_cache: &query_cache,
            ignore_markers: None,
        };
        store.observers_for(event.kind()).observe(
            &handler_module,
            handler_module.heap().alloc(event),
            observe_opts,
        )?;
        handler_module
            .into_intents_on(frozen_heap)?
            .into_iter()
            .for_each(|intent| match intent {
                Intent::Find {
                    language,
                    query,
                    on_match,
                } => project_queries.push((language, query, on_match)),
                Intent::Observe { .. } => panic!("internal error: non-init observe"),
                Intent::Warn(irr) => irritations.push(irr),
            });
        project_queries
    };

    for file in &files {
        let Some(language) = file.language() else {
            if log_enabled!(log::Level::Info) {
                info!("skipping {}", file.path());
            }
            continue;
        };

        let file_queries = {
            let mut file_queries = Vec::with_capacity(store.file_queries_hint());
            let path = file.path().pretty_path.dupe();

            let event = OpenFileEvent::new(path);
            let handler_module = HandlerModule::new();
            let observe_opts = ObserveOptions {
                action: Action::Vexing(event.kind()),
                query_cache: &query_cache,
                ignore_markers: None,
            };
            store.observers_for(event.kind()).observe(
                &handler_module,
                handler_module.heap().alloc(event),
                observe_opts,
            )?;
            handler_module
                .into_intents_on(frozen_heap)?
                .into_iter()
                .for_each(|intent| match intent {
                    Intent::Find {
                        language,
                        query,
                        on_match,
                    } => file_queries.push((language, query, on_match)),
                    Intent::Observe { .. } => panic!("internal error: non-init observe"),
                    Intent::Warn(irr) => irritations.push(irr.clone()),
                });
            file_queries
        };

        if project_queries
            .iter()
            .chain(file_queries.iter())
            .all(|(l, _, _)| *l != language)
        {
            continue; // No need to parse, the user will never search this.
        }
        let parsed_file = file.parse()?;
        let ignore_markers = parsed_file.ignore_markers()?;
        project_queries
            .iter()
            .chain(file_queries.iter())
            .filter(|(l, _, _)| *l == language)
            .try_for_each(|(_, query, on_match)| {
                QueryCursor::new()
                    .matches(
                        query,
                        parsed_file.tree.root_node(),
                        parsed_file.content.as_bytes(),
                    )
                    .try_for_each(|qmatch| {
                        let handler_module = HandlerModule::new();
                        let event = {
                            let path = parsed_file.path.pretty_path.dupe();
                            let captures = QueryCaptures::new(
                                query,
                                qmatch,
                                &parsed_file,
                                handler_module.heap(),
                            );
                            handler_module.heap().alloc(MatchEvent::new(path, captures))
                        };
                        let observe_opts = ObserveOptions {
                            action: Action::Vexing(EventKind::Match),
                            query_cache: &query_cache,
                            ignore_markers: Some(&ignore_markers),
                        };
                        on_match.observe(&handler_module, event, observe_opts)?;
                        handler_module
                            .into_intents_on(frozen_heap)?
                            .into_iter()
                            .for_each(|intent| match intent {
                                Intent::Find { .. } => {
                                    panic!("internal error: find intended during find")
                                }
                                Intent::Observe { .. } => {
                                    panic!("internal error: non-init observe")
                                }
                                Intent::Warn(irr) => irritations.push(irr),
                            });

                        Ok::<_, Error>(())
                    })
            })?;
    }

    irritations.sort();
    if let MaxProblems::Limited(max) = max_problems {
        let max = max as usize;
        if max < irritations.len() {
            irritations.truncate(max);
        }
    }
    Ok(RunData {
        irritations,
        num_files_scanned: files.len(),
    })
}

fn walkdir(
    ctx: &Context,
    path: &Utf8Path,
    ignores: &[FilePattern],
    allows: &[FilePattern],
    paths: &mut Vec<Utf8PathBuf>,
) -> Result<()> {
    if log_enabled!(log::Level::Trace) {
        trace!("walking {path}");
    }
    let entries = fs::read_dir(path).map_err(|cause| Error::IO {
        path: PrettyPath::new(path),
        action: IOAction::Read,
        cause,
    })?;
    for entry in entries {
        let entry = entry.map_err(|cause| Error::IO {
            path: PrettyPath::new(path),
            action: IOAction::Read,
            cause,
        })?;
        let entry_path = Utf8PathBuf::try_from(entry.path())?;
        let metadata = fs::symlink_metadata(&entry_path).map_err(|cause| Error::IO {
            path: PrettyPath::new(&entry_path),
            action: IOAction::Read,
            cause,
        })?;
        let is_dir = metadata.is_dir();

        let project_relative_path =
            Utf8Path::new(&entry_path.as_str()[ctx.project_root.as_str().len()..]);
        if !allows.iter().any(|p| p.matches(project_relative_path)) {
            let hidden = project_relative_path
                .file_name()
                .is_some_and(|name| name.starts_with('.'));
            if hidden || ignores.iter().any(|p| p.matches(project_relative_path)) {
                if log_enabled!(log::Level::Info) {
                    let dir_marker = if is_dir { "/" } else { "" };
                    info!("ignoring {project_relative_path}{dir_marker}");
                }
                continue;
            }
        }

        if metadata.is_symlink() {
            if log_enabled!(log::Level::Info) {
                let symlink_path = entry_path.strip_prefix(ctx.project_root.as_ref())?;
                info!("ignoring /{symlink_path} (symlink)");
            }
        } else if is_dir {
            walkdir(ctx, &entry_path, ignores, allows, paths)?;
        } else if metadata.is_file() {
            paths.push(entry_path);
        } else {
            panic!("unreachable");
        }
    }

    Ok(())
}

fn init(init_args: InitCmd) -> Result<()> {
    let cwd = Utf8PathBuf::try_from(env::current_dir().map_err(|cause| Error::IO {
        path: PrettyPath::new(Utf8Path::new(".")),
        action: IOAction::Read,
        cause,
    })?)?;
    Context::init(cwd, init_args.force)?;
    let queries_dir = Context::acquire()?.manifest.queries_dir;
    printdoc!(
        "
            {}: vex initialised
            now add style rules in ./{}/
            for an example, open ./{}/{EXAMPLE_VEX_FILE}
        ",
        "success".if_supports_color(Stream::Stdout, |text| text.style(*SUCCESS_STYLE)),
        queries_dir.as_str(),
        queries_dir.as_str(),
    );
    Ok(())
}

#[cfg(test)]
mod test {
    use indoc::indoc;
    use insta::assert_yaml_snapshot;
    use joinery::JoinableIterator;

    use crate::vextest::VexTest;

    use super::*;

    #[test]
    fn max_problems() {
        const MAX: u32 = 47;
        let irritations = VexTest::new("max-problems")
            .with_max_problems(MaxProblems::Limited(MAX))
            .with_scriptlet(
                "vexes/var.star",
                indoc! {r#"
                    def init():
                        vex.observe('open_project', on_open_project)

                    def on_open_project(event):
                        vex.search(
                            'rust',
                            '(integer_literal) @num',
                            on_match,
                        )

                    def on_match(event):
                        vex.warn('oh no a number!', at=(event.captures['num'], 'num'))
                "#},
            )
            .with_source_file(
                "src/main.rs",
                indoc! {r#"
                    fn main() {
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        let x = 1 + 2 + 3 + 4 + 5 + 6 + 8 + 9 + 10;
                        println!("{x}");
                    }
                "#},
            )
            .try_run()
            .unwrap()
            .into_irritations();
        assert_eq!(irritations.len(), MAX as usize);
    }

    #[test]
    fn readme() {
        // Dumb hacky test to serve until mdbook docs are made and tested.
        let collate_snippets = |language| {
            include_str!("../README.md")
                .lines()
                .scan(false, |collate_starlark, line| {
                    Some(if let Some(stripped) = line.strip_prefix("```") {
                        *collate_starlark = stripped.starts_with(language);
                        None
                    } else if *collate_starlark {
                        Some(line)
                    } else {
                        None
                    })
                })
                .flatten()
                .join_with("\n")
                .to_string()
        };
        let collated_starlark_snippets = collate_snippets("python");
        let collated_rust_snippets = collate_snippets("rust");
        let irritations = VexTest::new("README-snippets")
            .with_scriptlet("vexes/test.star", collated_starlark_snippets)
            .with_source_file("src/main.rs", collated_rust_snippets)
            .try_run()
            .unwrap()
            .into_irritations()
            .into_iter()
            .map(|irr| irr.to_string())
            .collect::<Vec<_>>();
        assert_yaml_snapshot!(irritations);
    }
}
