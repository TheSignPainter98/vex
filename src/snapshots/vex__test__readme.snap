---
source: src/main.rs
expression: irritations
---
- "warning[test]: large operands should come later\n --> src/main.rs:2:5\n  |\n2 |     123456 + 1\n  |     ------ number too large\n  |\n  = info: large numbers can hide smaller ones"
